zcash_primitives/src/merkle_tree.rs

```diff
@@ -444,48 +444,41 @@ impl<Node: Hashable> CommitmentTreeWitness<Node> {
         witness = &witness[1..];
 
         // Begin to construct the authentication path
-        let mut auth_path = vec![None; depth];
+        let iter = witness.chunks_exact(33);
+        witness = iter.remainder();
 
         // The vector works in reverse
-        for i in (0..depth).rev() {
-            // skip length of inner vector
-            if witness[0] != 32 {
-                // the length of a pedersen hash
-                return Err(());
-            }
-            witness = &witness[1..];
-
-            // Grab the sibling node at this depth in the tree
-            let mut sibling = [0u8; 32];
-            sibling.copy_from_slice(&witness[0..32]);
-            witness = &witness[32..];
-
-            // Sibling node should be an element of Fr
-            let sibling = match Node::read(&sibling[..]) {
-                Ok(p) => p,
-                Err(_) => return Err(()),
-            };
-
-            // Set the value in the auth path; we put false here
-            // for now (signifying the position bit) which we'll
-            // fill in later.
-            auth_path[i] = Some((sibling, false));
+        let mut auth_path = iter
+            .rev()
+            .map(|bytes| {
+                // Length of inner vector should be the length of a Pedersen hash
+                if bytes[0] == 32 {
+                    // Sibling node should be an element of Fr
+                    Node::read(&bytes[1..])
+                        .map(|sibling| {
+                            // Set the value in the auth path; we put false here
+                            // for now (signifying the position bit) which we'll
+                            // fill in later.
+                            (sibling, false)
+                        })
+                        .map_err(|_| ())
+                } else {
+                    Err(())
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        if auth_path.len() != depth {
+            return Err(());
         }
 
         // Read the position from the witness
-        let position = match witness.read_u64::<LittleEndian>() {
-            Ok(pos) => pos,
-            Err(_) => return Err(()),
-        };
+        let position = witness.read_u64::<LittleEndian>().map_err(|_| ())?;
```

Comment:
Maybe `if witness.len() != 8 { return Err(()) }`?

Reply:
This case is already covered: `ReadBytesExt::read_u64` returns the same errors are `Read::read_exact`, including an error if there are not enough bytes in the reader to fill the target.